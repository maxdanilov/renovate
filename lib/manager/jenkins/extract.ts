import * as datasourceJenkins from '../../datasource/jenkins';
import { logger } from '../../logger';
import * as dockerVersioning from '../../versioning/docker';
import { PackageDependency, PackageFile } from '../common';

export function extractPackageFile(content: string): PackageFile | null {
  logger.trace('jenkins.extractPackageFile()');
  const deps: PackageDependency[] = [];

  for (const line of content.split('\n')) {
    const match = /^\s*([\d\w-]+):([^#\s]+).*$/.exec(line);

    if (match) {
      const dep: PackageDependency = {
        datasource: datasourceJenkins.id,
        depName: match[1],
        currentValue: match[2],
        versioning: dockerVersioning.id,
      };

      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
        },
        'Jenkins plugin'
      );
      deps.push(dep);
    }
  }

  if (!deps.length) {
    return null;
  }
  return { deps };
}
