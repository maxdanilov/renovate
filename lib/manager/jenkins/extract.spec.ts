import { readFileSync } from 'fs';
import { extractPackageFile } from './extract';

const pluginsFile = readFileSync(
  'lib/manager/jenkins/__fixtures__/plugins.txt',
  'utf8'
);

const pluginsEmptyFile = readFileSync(
  'lib/manager/jenkins/__fixtures__/empty.txt',
  'utf8'
);

describe('lib/manager/jenkins/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      expect(extractPackageFile(pluginsEmptyFile)).toBeNull();
    });

    it('extracts multiple image lines', () => {
      const res = extractPackageFile(pluginsFile);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(4);
    });
  });
});
