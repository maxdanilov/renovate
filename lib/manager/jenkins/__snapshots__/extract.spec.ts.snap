// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/jenkins/extract extractPackageFile() extracts multiple image lines 1`] = `
Array [
  Object {
    "currentValue": "1.2.3",
    "datasource": "jenkins",
    "depName": "email-ext",
    "versioning": "docker",
  },
  Object {
    "currentValue": "4.4.10-2.0",
    "datasource": "jenkins",
    "depName": "apache-httpcomponents-client-4-api",
    "versioning": "docker",
  },
  Object {
    "currentValue": "1.2",
    "datasource": "jenkins",
    "depName": "authentication-tokens",
    "versioning": "docker",
  },
  Object {
    "currentValue": "1.21.0",
    "datasource": "jenkins",
    "depName": "blueocean",
    "versioning": "docker",
  },
]
`;
